google/cloud/spanner/internal/connection_impl.cc

```diff
@@ -706,20 +707,33 @@ StatusOr<CommitResult> ConnectionImpl::CommitImpl(
   for (auto&& m : params.mutations) {
     *request.add_mutations() = std::move(m).as_proto();
   }
-  bool is_idempotent = false;
-  if (s.has_single_use()) {
-    *request.mutable_single_use_transaction() = s.single_use();
-  } else if (s.has_begin()) {
-    *request.mutable_single_use_transaction() = s.begin();
-  } else {
-    request.set_transaction_id(s.id());
-    is_idempotent = true;
+
+  if (s.selector_case() != spanner_proto::TransactionSelector::kId) {
+    spanner_proto::BeginTransactionRequest begin;
+    begin.set_session(session->session_name());
+    *begin.mutable_options() = s.begin();
```

Comment:
Even though it doesn't make sense to pass a single-use/read-only transaction to `Commit()`, it also doesn't make sense to pass any read-only transaction, yet we support the latter and not the former.  So, we might as well look at `s.has_single_use()` and use its `TransactionOptions`, rather than using the default `TransactionOptions`, and let the chips fall where they may.

Reply:
I wouldn't say that we "support" read-only transactions. Any caller who passes in a read-only transaction to `Commit()` will get an error. I believe that's true before and after this PR.

To call `Commit()` correctly, callers must pass in a read-write transaction. RW transactions have all their options in the "begin" field of the selector, so that's why I think we only need to copy that field from the selector into the `mutable_options` in the request.

I don't think we should write extra code to deal with things like single-use RO transactions when callers are not allowed to pass them, and if they did, it's guaranteed to fail.

Reply:
Discussed this offline. I changed the code to consider the case where the input selector was a single_use transaction as @devbww was suggesting. We know this will give an error, but it's what the user asked for.